src/scheduler.rs

```diff
@@ -99,16 +116,53 @@ impl SchedulerState {
     pub(crate) async fn pause<'a>(&'_ self, context: Context) -> IoPausedGuard {
         {
             let mut inner = self.inner.write().await;
-            inner.paused += 1;
-            Self::do_stop(inner, &context).await;
+            match *inner {
+                InnerSchedulerState::Started(_) => {
+                    let new_state = InnerSchedulerState::Paused {
+                        started: true,
+                        pause_guards_count: NonZeroUsize::new(1).unwrap(),
+                    };
+                    Self::do_stop(inner, &context, new_state).await;
+                }
+                InnerSchedulerState::Stopped => {
+                    *inner = InnerSchedulerState::Paused {
+                        started: false,
+                        pause_guards_count: NonZeroUsize::new(1).unwrap(),
+                    };
+                }
+                InnerSchedulerState::Paused {
+                    ref mut pause_guards_count,
+                    ..
+                } => *pause_guards_count = pause_guards_count.checked_add(1).unwrap(),
+            }
         }
+
         let (tx, rx) = oneshot::channel();
         tokio::spawn(async move {
             rx.await.ok();
             let mut inner = context.scheduler.inner.write().await;
-            inner.paused -= 1;
-            if inner.paused == 0 && inner.started && inner.scheduler.is_none() {
-                SchedulerState::do_start(inner, context.clone()).await;
+            match *inner {
+                InnerSchedulerState::Started(_) => {
+                    warn!(&context, "IoPausedGuard resume: started instead of paused");
+                }
+                InnerSchedulerState::Stopped => {
+                    warn!(&context, "IoPausedGuard resume: stopped instead of paused");
+                }
+                InnerSchedulerState::Paused {
+                    ref started,
+                    ref mut pause_guards_count,
+                } => {
+                    if *pause_guards_count == NonZeroUsize::new(1).unwrap() {
```

Comment:
Yeah, I know NonZero* types do optimise storage (though was really disappointed to see it doesn't do it in this case, IIUC it totally could put the bool and NonZeroUsize in a single usize-sized word - not that this matters for us though).  I still think it's a nice newtype, though it could be nicer to use indeed.

I'm a bit undecided.  I think the current version is reasonable, but it does have unwrap calls.  The signature has to change to return a `Result<IoPausedGuard>` in any case to be safe.  If you both vote for usize + counting from zero I'm happy to change it.